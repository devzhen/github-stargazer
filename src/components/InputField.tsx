import React from 'react';
import { isNil, isEmpty } from 'ramda';
import { KeyboardTypeOptions, ReturnKeyTypeOptions } from 'react-native';
import styled from 'styled-components/native';

import { IS_ANDROID } from '../helpers/deviceInfoHelper';

interface InputFieldProps {
  offsetTop?: number;
  offsetBottom?: number;
  offsetLeft?: number;
  offsetRight?: number;
  fontSize?: number;
  onChangeText: (value: string) => void;
  value: string;
  placeholder?: string;
  keyboardType: KeyboardTypeOptions;
  returnKeyType: ReturnKeyTypeOptions;
  secure?: boolean;
  onSubmitEditing?: any;
  color?: string;
  innerRef: any;
  inputAccessoryViewID?: any;
  error?: boolean;
  inputHeight?: number;
}

const InputField = ({
  offsetTop,
  offsetLeft,
  offsetRight,
  offsetBottom,
  fontSize,
  onChangeText,
  value,
  placeholder,
  keyboardType,
  returnKeyType,
  secure,
  onSubmitEditing,
  color,
  innerRef,
  inputAccessoryViewID,
  error,
  inputHeight,
}: InputFieldProps) => (
  <Container
    offsetTop={offsetTop}
    offsetLeft={offsetLeft}
    offsetRight={offsetRight}
    offsetBottom={offsetBottom}
  >
    <StyledInput
      inputHeight={inputHeight}
      ref={innerRef}
      withOffset={!isNil(value) && !isEmpty(value)}
      fontSize={fontSize}
      onChangeText={onChangeText}
      value={value}
      placeholderTextColor="#9a9a9a"
      placeholder={placeholder}
      keyboardType={keyboardType}
      returnKeyType={returnKeyType}
      underlineColorAndroid="transparent"
      secureTextEntry={secure}
      onSubmitEditing={onSubmitEditing}
      autoCapitalize="none"
      autoCorrect={false}
      color={color}
      inputAccessoryViewID={inputAccessoryViewID}
      isEmpty={isEmpty(value) || isNil(value)}
    />
    {!isNil(value) && !isEmpty(value) && (
      <PlaceholderWrapper>
        <Placeholder error={error}>{placeholder}</Placeholder>
      </PlaceholderWrapper>
    )}
  </Container>
);

InputField.defaultProps = {
  offsetTop: 0,
  offsetBottom: 0,
  offsetLeft: 0,
  offsetRight: 0,
  fontSize: '17px',
  placeholder: '',
  keyboardType: 'default',
  secure: false,
  onSubmitEditing: () => {},
  returnKeyType: 'next',
  innerRef: () => {},
  autoFocus: false,
  centeringText: false,
  width: '100%',
  color: '#383838',
  inputHeight: 45,
};

const Container = styled.View<{
  offsetTop?: number;
  offsetBottom?: number;
  offsetLeft?: number;
  offsetRight?: number;
}>`
  margin-top: ${(props) => props.offsetTop || 0}px;
  margin-bottom: ${(props) => props.offsetBottom || 0}px;
  margin-left: ${(props) => props.offsetLeft || 0}px;
  margin-right: ${(props) => props.offsetRight || 0}px;
`;

const StyledInput = styled.TextInput<{
  fontSize?: number;
  color?: string;
  withOffset: boolean;
  isEmpty: boolean;
  inputHeight?: number;
}>`
  font-size: ${(props) => props.fontSize};
  color: ${(props) => props.color};
  width: 100%;
  height: ${(props) => props.inputHeight || 45}px;
  padding-left: 19px;
  background-color: ${(props) => props.theme.colors.porcelain};
  border-color: ${(props) => props.theme.colors.gray};
  border-width: 1px;
  border-radius: 4px;
  ${(props) => props.withOffset && 'padding-top: 14px'};
  ${(props) => IS_ANDROID && !props.isEmpty && 'padding-bottom: 0'};
`;

const PlaceholderWrapper = styled.View`
  position: absolute;
  top: 2px;
  left: 0;
  width: 100%;
  padding-left: 19px;
`;

const Placeholder = styled.Text<{ error?: boolean }>`
  font-size: 12px;
  color: ${(props) => props.theme.colors.orange};
  font-weight: 600;
  ${(props) => props.error && `color: ${props.theme.colors.error}`};
`;

export default InputField;
