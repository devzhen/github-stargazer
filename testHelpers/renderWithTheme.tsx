import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import 'jest-styled-components';

import theme from '@theme/index';

const renderWithTheme = (component: any) =>
  renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

export default renderWithTheme;
