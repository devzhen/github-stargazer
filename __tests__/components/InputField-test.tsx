import React from 'react';
import 'jest-styled-components';

import InputField from '@components/InputField';

// @ts-ignore
import renderWithTheme from '@testHelpers/renderWithTheme';

describe('InputField common component', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(
      <InputField value="test" onChangeText={() => {}} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
