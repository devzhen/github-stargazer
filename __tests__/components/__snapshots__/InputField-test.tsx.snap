// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField common component renders correctly 1`] = `
<View
  offsetBottom={0}
  offsetLeft={0}
  offsetRight={0}
  offsetTop={0}
  style={
    Array [
      Object {
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    autoCorrect={false}
    color="#383838"
    fontSize="17px"
    inputHeight={45}
    isEmpty={false}
    keyboardType="default"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder=""
    placeholderTextColor="#9a9a9a"
    rejectResponderTermination={true}
    returnKeyType="next"
    secureTextEntry={false}
    style={
      Array [
        Object {
          "backgroundColor": "#f6f7f8",
          "borderColor": "#d3d3d3",
          "borderRadius": 4,
          "borderWidth": 1,
          "color": "#383838",
          "fontSize": 17,
          "height": 45,
          "paddingLeft": 19,
          "paddingTop": 14,
          "width": "100%",
        },
      ]
    }
    underlineColorAndroid="transparent"
    value="test"
    withOffset={true}
  />
  <View
    style={
      Array [
        Object {
          "left": 0,
          "paddingLeft": 19,
          "position": "absolute",
          "top": 2,
          "width": "100%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#eb9e38",
            "fontSize": 12,
            "fontWeight": "600",
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
